### PR TITLE
Feature/send payment

### PR DESCRIPTION
**Pull request by roberthein**
Implemented feature to send payment to an external payment address.

 img_3256

### CODE REVIEW

File: Toshi/Controllers/Payments/PaymentAddressController.swift [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    public func scannerViewController(_ controller: ScannerViewController, didScanResult result: String) {
+        
+        guard let url = URL(string: result) as URL? else { return }
+        let path = url.path
+        
+        if path.hasPrefix("/add") {
+            let username = result.replacingOccurrences(of: QRCodeController.addUsernameBasePath, with: "")
+            let contactName = TokenUser.name(from: username)
+            
+            IDAPIClient.shared.retrieveContact(username: contactName) { contact in
```
Comment (heumn) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: I think you forgot a [weak self] here? 🤔  😄

File: Toshi/Controllers/Payments/PaymentRequestController.swift [hunk 2]
```diff
@@ -20,28 +20,19 @@ protocol PaymentRequestControllerDelegate: class {
...
-    }()
-
-    lazy var cancelBarButton: UIBarButtonItem = {
-        let item = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancelRequest))
-
-        return item
-    }()
-
+    lazy var continueBarButton = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(sendRequest))
+    lazy var cancelBarButton = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancelRequest))
```
Comment (heumn) [inline Toshi/Controllers/Payments/PaymentRequestController.swift]: Nice 👌

File: Toshi/Controllers/Payments/PaymentSendController.swift [hunk 3]
```diff
@@ -16,39 +16,37 @@
...
-    }()
-
-    lazy var cancelBarButton: UIBarButtonItem = {
-        let item = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancel))
-
-        return item
-    }()
+    
+    lazy var continueBarButton = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(send))
+    lazy var cancelBarButton = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancel))
```
Comment (heumn) [inline Toshi/Controllers/Payments/PaymentSendController.swift]: 👍

File: Toshi/Controllers/PaymentNavigationController.swift [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import UIKit
+
+public class PaymentNavigationController: UINavigationController {
```
Comment (heumn) [inline Toshi/Controllers/PaymentNavigationController.swift]: Maybe consider locking down classes the classes with `private` or `fileprivate`. High safety is a priority for me. Mostly because I have made so many mistakes and now build walls around my shortcomings. One of my typical ones has been the quick-fix route of subclassing something to fix "that one little thing I need / want to adjust". Setting up the barrier by "nope, this class is not designed for further subclassing" has been a nice guidance for me (and other coming after me?)

File: Toshi/Controllers/Payments/Payment.swift [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+let Payment = PaymentManager.shared
+
+class PaymentManager {
+    
+    static let shared: PaymentManager = {
+        return PaymentManager()
+    }()
+    
+    typealias SuccessCompletion = ((UIAlertAction) -> Void)
```
Comment (heumn) [inline Toshi/Controllers/Payments/Payment.swift]: Nice use of typealias! 👍

File: Toshi/Controllers/Payments/PaymentAddressController.swift [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    }()
+    
+    fileprivate lazy var addressInputView: PaymentAddressInputView = {
+        let view = PaymentAddressInputView()
+        view.delegate = self
+        
+        return view
+    }()
+    
+    lazy var sendBarButton = UIBarButtonItem(title: Localized("payment_send_button"), style: .plain, target: self, action: #selector(send))
```
Comment (heumn) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: This is another one of those "hey, Nikolai could arrive here in 3 months, see this property from the outside and think it is perfectly fine to grab and modify it". I would slap a `fileprivate" on that and say "hey man, no, this it not for you to touch!"

File: Toshi/Controllers/Payments/PaymentAddressController.swift [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        valueLabel.left(to: view, offset: 16)
+        valueLabel.right(to: view, offset: -16)
+        
+        descriptionLabel.topToBottom(of: valueLabel, offset: 10)
+        descriptionLabel.left(to: view, offset: 16)
+        descriptionLabel.right(to: view, offset: -16)
+        
+        addressInputView.topToBottom(of: descriptionLabel, offset: 40)
+        addressInputView.left(to: view)
+        addressInputView.right(to: view)
```
Comment (heumn) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: ☝️ my OCD loves you for sorting these like that <3

File: Toshi/Controllers/Payments/PaymentAddressController.swift [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        addressInputView.addressTextField.resignFirstResponder()
+    }
+}
+
+extension PaymentAddressController: PaymentAddressInputDelegate {
+
+    func showQRScanner() {
+        let scannerController = ScannerController(instructions: Localized("payment_qr_scanner_instructions"), types: [.qrCode])
+        scannerController.delegate = self
+        navigationController?.pushViewController(scannerController, animated: true)
```
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: i guess scanner is to be presented modally right ?

File: Toshi/Controllers/Payments/PaymentAddressController.swift [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    public func scannerViewControllerDidCancel(_ controller: ScannerViewController) {
+        controller.navigationController?.popViewController(animated: true)
+    }
+    
+    public func scannerViewController(_ controller: ScannerViewController, didScanResult result: String) {
+        
+        guard let url = URL(string: result) as URL? else { return }
+        let path = url.path
+        
+        if path.hasPrefix("/add") {
```
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: should be `/add` here ?

File: Toshi/Controllers/Payments/PaymentAddressInputView.swift [hunk 10]
```diff
@@ -0,0 +1,100 @@
+import Foundation
+import UIKit
+import TinyConstraints
+
+protocol PaymentAddressInputDelegate: class {
+    func showQRScanner()
+    func send()
```
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentAddressInputView.swift]: naming is not good i think, lets use proper delegate namings ? with all needed parameters

File: Toshi/Controllers/Payments/PaymentRequestController.swift (hunk 2 above)
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentRequestController.swift]: lets call buttons action in a better way ? `sendRequest / cancelRequest` sound like simple functions

File: Toshi/Controllers/Payments/PaymentRequestController.swift [hunk 11]
```diff
@@ -20,28 +20,19 @@ protocol PaymentRequestControllerDelegate: class {
...
-        return item
-    }()
-
-    lazy var cancelBarButton: UIBarButtonItem = {
-        let item = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancelRequest))
-
-        return item
-    }()
-
+    lazy var continueBarButton = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(sendRequest))
```
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentRequestController.swift]: also private please

File: Toshi/Controllers/Payments/PaymentSendController.swift [hunk 12]
```diff
@@ -16,39 +16,37 @@
 import UIKit
 
 protocol PaymentSendControllerDelegate: class {
-    func paymentSendControllerDidFinish(valueInWei: NSDecimalNumber?)
+    func paymentSendControllerCanceled()
```
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentSendController.swift]: please, proper delegate namings ?
`paymentSendControllerDidCancel(_ controller: PaymentSendController)`

File: Toshi/Controllers/Payments/PaymentSendController.swift [hunk 13]
```diff
@@ -16,39 +16,37 @@
...
+    func paymentSendControllerFinished(with valueInWei: NSDecimalNumber?, for controller: PaymentSendController)
```
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentSendController.swift]: `func paymentSendControllerDidFinish(_ controller: PaymentSendController, with valueInWei: NSDecimalNumber?)`

File: Toshi/Controllers/Payments/PaymentSendController.swift [hunk 14]
```diff
@@ -16,39 +16,37 @@
...
-        return item
-    }()
-
-    lazy var cancelBarButton: UIBarButtonItem = {
-        let item = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancel))
-
-        return item
-    }()
+    
+    lazy var continueBarButton = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(send))
```
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentSendController.swift]: private ?

Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentSendController.swift]: and actions names - `didPressedContinueButton` ?

File: Toshi/Controllers/Payments/PaymentSendController.swift [hunk 15]
```diff
@@ -16,39 +16,37 @@
...
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        
+        navigationItem.backBarButtonItem = UIBarButtonItem.back
     }
 
     func cancel() {
-        delegate?.paymentSendControllerDidFinish(valueInWei: nil)
+        delegate?.paymentSendControllerCanceled()
```
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentSendController.swift]: yeah, actually u dnt need a delegate call for this .

File: Toshi/Controllers/ScannerController.swift [hunk 16]
```diff
@@ -75,7 +87,7 @@ extension ScannerController: PaymentPresentable {
 
     func paymentApproved(with parameters: [String: Any], userInfo: UserInfo) {
         isStatusBarHidden = false
-        guard userInfo.isLocal == false else {
+        guard !userInfo.isLocal else {
```
Comment (yuliaveres) [inline Toshi/Controllers/ScannerController.swift]: doesn't it collides a bit with force unwrap ? 😄  for me it does. I would just suggest to use `== false` for this to avoid confusions ?

File: Toshi/MessagesUI/Objects/Extensions.swift [hunk 17]
```diff
@@ -123,6 +123,15 @@ extension UIImageView {
     }
 }
 
+extension UIBarButtonItem {
+    
+    static var back: UIBarButtonItem {
+        let item = UIBarButtonItem()
+        item.title = "Back"
+        return item
```
Comment (yuliaveres) [inline Toshi/MessagesUI/Objects/Extensions.swift]: new line ?

File: Toshi/Controllers/PaymentNavigationController.swift (hunk 4 above)
Comment (heumn) [inline Toshi/Controllers/PaymentNavigationController.swift]: Sorry, yeah. My bad. I meant of course final. What is a `fileprivate` class 😂

File: Toshi/Controllers/Payments/PaymentRequestController.swift (hunk 2 above)
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentRequestController.swift]: Can you make a suggestion? When you request a name change it's good to provide your suggestion, because else I'm just guessing. 😉

Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentRequestController.swift]: sure ! :) cancelButtonDidPress / didPressCancelButton ?

File: Toshi/Controllers/Payments/PaymentSendController.swift (hunk 3 above)
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentSendController.swift]: sendButtonDidPress / didPressSendButton ? .. same for all of them

File: Toshi/Controllers/Payments/PaymentAddressController.swift (hunk 1 above)
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: Yes! I want a SwiftLint or Danger rule for this, because there are a lot of cases like this in the app.

File: Toshi/Controllers/PaymentNavigationController.swift (hunk 4 above)
Comment (roberthein) [inline Toshi/Controllers/PaymentNavigationController.swift]: Ok! I'll try to add this to Danger or SwiftLint as well.

File: Toshi/Controllers/Payments/Payment.swift (hunk 5 above)
Comment (roberthein) [inline Toshi/Controllers/Payments/Payment.swift]: 😏

File: Toshi/Controllers/Payments/PaymentAddressInputView.swift (hunk 10 above)
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentAddressInputView.swift]: `func paymentAddressInputViewDidRequestShowScanner(_ view: PaymentAddressInputView)`
`func paymentAddressInputViewDidSend(_ view: PaymentAddressInputView)`

File: Toshi/Controllers/Payments/PaymentAddressController.swift (hunk 6 above)
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: Made it `private` instead. Now stop touching it!

File: Toshi/Controllers/Payments/PaymentAddressController.swift (hunk 7 above)
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: `top`, `left`, `bottom`, `right`?

File: Toshi/Controllers/Payments/PaymentAddressInputView.swift (hunk 10 above)
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentAddressInputView.swift]: I don't think we should repeat the class name in the function name because the `PaymentAddressInputDelegate` protocol can be implemented by other classes than the `PaymentAddressInputView`, context can be derived from the extension. 
Why do I need to add a parameter that is not being used and also limits the protocol for only being used with the `PaymentAddressInputView` class.
I can agree to better function names than the current ones, but not this.

File: Toshi/Controllers/Payments/PaymentAddressController.swift (hunk 9 above)
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: I think so, why do you doubt this?

File: Toshi/Controllers/Payments/PaymentAddressController.swift [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+    func showQRScanner() {
+        let scannerController = ScannerController(instructions: Localized("payment_qr_scanner_instructions"), types: [.qrCode])
+        scannerController.delegate = self
+        navigationController?.pushViewController(scannerController, animated: true)
+    }
+    
+    func send() {
+        guard let paymentAddress = addressInputView.addressTextField.text else { return }
+        
+        Payment.send(valueInWei, to: paymentAddress) { [weak self] _ in
```
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: I believe `Payment.send(valueInWei, to: paymentAddress)` reads better than `Payment.shared.send(valueInWei, to: paymentAddress)`..

File: Toshi/Controllers/Payments/PaymentAddressController.swift (hunk 8 above)
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: Might be better indeed. ✌️

File: Toshi/Controllers/Payments/PaymentAddressController.swift [hunk 19]
```diff
@@ -0,0 +1,135 @@
+import Foundation
+import UIKit
+import TinyConstraints
+import CameraScanner
+
+class PaymentAddressController: UIViewController {
+    
+    let valueInWei: NSDecimalNumber
```
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: I guess so! 👍

File: Toshi/Controllers/Payments/Payment.swift [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+            "to": paymentAddress,
+            "value": valueInWei.toHexString
+        ]
+        
+        EthereumAPIClient.shared.createUnsignedTransaction(parameters: parameters) { transaction, error in
+            
+            guard let transaction = transaction as String? else {
+                
+                if let error = error {
+                    self.showPaymentFailedMessage(for: error.localizedDescription)
```
Comment (roberthein) [inline Toshi/Controllers/Payments/Payment.swift]: I think so too, but that's going to be a different PR.

File: Toshi/Controllers/Payments/Payment.swift [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+        
+        let parameters: [String: Any] = [
+            "from": Cereal.shared.paymentAddress,
+            "to": paymentAddress,
+            "value": valueInWei.toHexString
+        ]
+        
+        EthereumAPIClient.shared.createUnsignedTransaction(parameters: parameters) { transaction, error in
+            
+            guard let transaction = transaction as String? else {
```
Comment (roberthein) [inline Toshi/Controllers/Payments/Payment.swift]: Fixed, I hope to make a guard for this with SwiftLint.

File: Toshi/Controllers/Payments/Payment.swift [hunk 22]
```diff
@@ -0,0 +1,81 @@
+import Foundation
+import UIKit
+
+let Payment = PaymentManager.shared
+
+class PaymentManager {
+    
+    static let shared: PaymentManager = {
```
Comment (roberthein) [inline Toshi/Controllers/Payments/Payment.swift]: ok

File: Toshi/Controllers/Payments/Payment.swift [hunk 23]
```diff
@@ -0,0 +1,81 @@
+import Foundation
+import UIKit
+
+let Payment = PaymentManager.shared
```
Comment (roberthein) [inline Toshi/Controllers/Payments/Payment.swift]: Because I want to say `payment send` instead of `payment manager shared send` 🤷‍♂️

File: Toshi/Controllers/Messaging/ChatController.swift [hunk 24]
```diff
@@ -883,22 +886,24 @@ extension ChatController: ChatsFloatingHeaderViewDelegate {
...
-
-    func paymentSendControllerDidFinish(valueInWei: NSDecimalNumber?) {
-        defer {
-            self.dismiss(animated: true)
-        }
-
-        guard let valueInWei = valueInWei else { return }
+    
+    func paymentSendControllerCanceled() {
+        dismiss(animated: true)
```
Comment (roberthein) [inline Toshi/Controllers/Messaging/ChatController.swift]: Nice one! Removed it!

File: Toshi/Controllers/Messaging/ChatController.swift [hunk 25]
```diff
@@ -872,9 +872,12 @@ extension ChatController: ImagePickerDelegate {
 extension ChatController: ChatsFloatingHeaderViewDelegate {
 
     func messagesFloatingView(_: ChatsFloatingHeaderView, didPressRequestButton _: UIButton) {
+        
         let paymentRequestController = PaymentRequestController()
         paymentRequestController.delegate = self
-        Navigator.presentModally(paymentRequestController)
+        
+        let navigationController = PaymentNavigationController(rootViewController: paymentRequestController)
```
Comment (roberthein) [inline Toshi/Controllers/Messaging/ChatController.swift]: Ok but different PR.

File: Toshi/Controllers/Payments/Payment.swift (hunk 23 above)
Comment (yuliaveres) [inline Toshi/Controllers/Payments/Payment.swift]: ok, as u wish 👍

File: Toshi/Controllers/Messaging/ChatController.swift (hunk 25 above)
Comment (roberthein) [inline Toshi/Controllers/Messaging/ChatController.swift]: The `PaymentNavigationController` has a different design than the others. I hope we can move to all-the-same-design later. I'll discuss this with Robert.

File: Toshi/Controllers/Favorites/ContactController.swift [hunk 26]
```diff
@@ -445,14 +446,14 @@ extension ContactController: RateUserControllerDelegate {
...
-        defer {
-            self.dismiss(animated: true)
-        }
-
-        guard let value = valueInWei else {
-            return
-        }
+    
+    func paymentSendControllerCanceled() {
+        dismiss(animated: true)
```
Comment (roberthein) [inline Toshi/Controllers/Favorites/ContactController.swift]: Removed it! ✨

File: Toshi/Controllers/Payments/PaymentSendController.swift (hunk 12 above)
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentSendController.swift]: I don't see why I should add the `PaymentSendController` when it's not being used, it just clutters the method.

File: Toshi/Controllers/Payments/PaymentSendController.swift (hunk 13 above)
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentSendController.swift]: I don't see why I should add the `PaymentSendController` when it's not being used, it just clutters the method.

File: Toshi/Controllers/Payments/PaymentSendController.swift (hunk 15 above)
Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentSendController.swift]: 👍

File: Toshi/Controllers/ScannerController.swift (hunk 16 above)
Comment (roberthein) [inline Toshi/Controllers/ScannerController.swift]: I don't agree, this should be fine. Do you also suggest using `guard userInfo.isLocal == true` instead of `guard userInfo.isLocal`?

File: Toshi/MessagesUI/Objects/Extensions.swift (hunk 17 above)
Comment (roberthein) [inline Toshi/MessagesUI/Objects/Extensions.swift]: ?

Comment (yuliaveres) [inline Toshi/MessagesUI/Objects/Extensions.swift]: before return ?

File: Toshi/Controllers/ScannerController.swift (hunk 16 above)
Comment (yuliaveres) [inline Toshi/Controllers/ScannerController.swift]: oh not sure , think isLocal is fine ? wdyt ?

File: Toshi/MessagesUI/Objects/Extensions.swift (hunk 17 above)
Comment (roberthein) [inline Toshi/MessagesUI/Objects/Extensions.swift]: done ;)

File: Toshi/Controllers/ScannerController.swift (hunk 16 above)
Comment (roberthein) [inline Toshi/Controllers/ScannerController.swift]: Both is fine `userInfo.isLocal` and `!userInfo.isLocal`.

File: Toshi/Controllers/Messaging/ChatController.swift (hunk 25 above)
Comment (yuliaveres) [inline Toshi/Controllers/Messaging/ChatController.swift]: that would be cool to have them similar, yep

File: Toshi/Controllers/Payments/Payment.swift (hunk 20 above)
Comment (yuliaveres) [inline Toshi/Controllers/Payments/Payment.swift]: 😿  ok

File: Toshi/Controllers/Payments/PaymentAddressController.swift (hunk 9 above)
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: cause add prefix is used for adding users :/ here the purpose to find, hmm.
@faymontage  can you approve the prefix is good here ?

File: Toshi/Controllers/ScannerController.swift (hunk 16 above)
Comment (yuliaveres) [inline Toshi/Controllers/ScannerController.swift]: ok

File: Toshi/Controllers/Payments/PaymentAddressController.swift (hunk 9 above)
Comment (faymontage) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: QRs with prefix "/add" are reserved for the intent of adding a contact and should not be processed here

File: Toshi/Controllers/Payments/PaymentAddressController.swift [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+
+extension PaymentAddressController: PaymentAddressInputDelegate {
+
+    func showQRScanner() {
+        let scannerController = ScannerController(instructions: Localized("payment_qr_scanner_instructions"), types: [.qrCode])
+        scannerController.delegate = self
+        navigationController?.pushViewController(scannerController, animated: true)
+    }
+    
+    func send() {
```
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentAddressController.swift]: what about this @roberthein ?

File: Toshi/Controllers/Payments/PaymentRequestController.swift (hunk 11 above)
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentRequestController.swift]: @roberthein  what about this ?

File: Toshi/Controllers/Payments/PaymentRequestController.swift (hunk 2 above)
Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentRequestController.swift]: @roberthein what about this ?

Comment (roberthein) [inline Toshi/Controllers/Payments/PaymentRequestController.swift]: I changed this right?

Comment (yuliaveres) [inline Toshi/Controllers/Payments/PaymentRequestController.swift]: oh Github doesnt show it, ok , sorry!
